File: packages/rook_sdk_apple_health/ios/Classes/Mapper/PhysicalSummaryMappers.swift
//
//  PhysicalSummaryMappers.swift
//  rook_sdk_apple_health
//
//  Created by Daniel Nolasco on 29/10/25.
//

import Foundation
import RookSDK

extension RookPhysicalSummary {
    func toProto() -> PhysicalSummaryProto {
        return PhysicalSummaryProto.with {
            $0.dateTime = dateTime.toProtoString()
            $0.sourceOfData = "Apple Health"
            $0.wasTheUserUnderPhysicalActivity = true
            $0.activeSeconds = activeSeconds.toInt32()
            $0.restSeconds = restSeconds.toInt32()
            $0.lowIntensitySeconds = lowIntensitySeconds.toInt32()
            $0.moderateIntensitySeconds = moderateIntensitySeconds.toInt32()
            $0.vigorousIntensitySeconds = vigorousIntensitySeconds.toInt32()
            $0.inactivitySeconds = inactivitySeconds.toInt32()
            $0.activityLevelGranularData = activityLevelGranularDataNumber?.map { $0.toGranularProto() } ?? []
            $0.continuousInactivePeriods = continuousInactivePeriodsNumber.toInt32()
            $0.caloriesNetIntakeKcal = caloriesExpenditureKilocalories.toDouble()
            $0.caloriesExpenditureKcal = caloriesExpenditureKilocalories.toDouble()
            $0.caloriesNetActiveKcal = caloriesNetActiveKilocalories.toDouble()
            $0.caloriesBasalMetabolicRateKcal = caloriesBasalMetabolicRateKilocalories.toDouble()
            $0.steps = stepsPerDayNumber.toInt32()
            $0.stepsGranularData = stepsGranularDataStepsPerHr?.map { $0.toStepsGranularProto() } ?? []
            $0.activeSteps = activeStepsPerDayNumber.toInt32()
            $0.activeStepsGranularData = activeStepsGranularDataStepsPerHr?
                .map { $0.toActiveStepsGranularProto() } ?? []
            $0.walkedDistanceMeters = walkedDistanceMeters.toDouble()
            $0.traveledDistanceMeters = traveledDistanceMeters.toDouble()
            $0.traveledDistanceGranularData = traveledDistanceGranularDataMeters?.map { $0.toGranularProto() } ?? []
            $0.floorsClimbed = floorsClimbedNumber.toDouble()
            $0.floorsClimbedGranularData = floorsClimbedGranularDataFloors?.map { $0.toGranularProto() } ?? []
            $0.elevationAvgAltitudeMeters = elevationAvgAltitudeMeters.toDouble()
            $0.elevationMinimumAltitudeMeters = elevationMinimumAltitudeMeters.toDouble()
            $0.elevationMaximumAltitudeMeters = elevationMaxAltitudeMeters.toDouble()
            $0.elevationLossActualAltitudeMeters = elevationLossActualAltitudeMeters.toDouble()
            $0.elevationGainActualAltitudeMeters = elevationGainActualAltitudeMeters.toDouble()
            $0.elevationPlannedGainMeters = elevationPlannedGainMeters.toDouble()
            $0.elevationGranularData = elevationGranularDataMeters?.map { $0.toGranularProto() } ?? []
            $0.swimmingNumStrokes = swimmingStrokesNumber.toInt32()
            $0.swimmingNumLaps = swimmingNumLapsNumber.toInt32()
            $0.swimmingPoolLengthMeters = swimmingPoolLengthMeters.toDouble()
            $0.swimmingTotalDistanceMeters = swimmingTotalDistanceMeters.toDouble()
            $0.swimmingDistanceGranularData = swimmingDistanceGranularDataMeters?.map { $0.toGranularProto() } ?? []
            $0.hrMaximumBpm = hrMaxBPM.toInt32()
            $0.hrMinimumBpm = hrMinimumBPM.toInt32()
            $0.hrAvgBpm = hrAvgBPM.toInt32()
            $0.hrRestingBpm = hrRestingBPM.toInt32()
            $0.hrGranularData = hrGranularDataBPM?.map { $0.toGranularProto() } ?? []
            $0.hrvAvgRmssd = hrvAvgRmssdNumber ?? DEFAULT_DOUBLE
            $0.hrvAvgSdnn = hrvAvgSdnnNumber ?? DEFAULT_DOUBLE
            $0.hrvSdnnGranularData = hrvSdnnGranularDataNumber?.map { $0.toGranularProto() } ?? []
            $0.hrvRmssdGranularData = hrvRmssdGranularDataNumber?.map { $0.toGranularProto() } ?? []
            $0.saturationAvgPercentage = saturationAvgPercentage.toDouble()
            $0.saturationGranularData = saturationGranularDataPercentage?.map { $0.toGranularProto() } ?? []
            $0.vo2MaxMlPerMinPerKg = vo2MaxMlPerMinPerKg ?? DEFAULT_DOUBLE
            $0.vo2GranularData = vo2GranularDataLiterPerMin?.map { $0.toGranularProto() } ?? []
            $0.stressAtRestDurationSeconds = stressAtRESTDurationSeconds.toInt32()
            $0.stressDurationSeconds = stressDurationSeconds.toInt32()
            $0.lowStressDurationSeconds = lowStressDurationSeconds.toInt32()
            $0.mediumStressDurationSeconds = mediumStressDurationSeconds.toInt32()
            $0.highStressDurationSeconds = highStressDurationSeconds.toInt32()
            $0.stressGranularData = stressGranularDataScoreNumber?.map { $0.toGranularProto() } ?? []
            $0.stressAvgLevel = stressAvgLevelNumber.toInt32()
            $0.stressMaximumLevel = stressMaxLevelNumber.toInt32()
        }
    }
}

File: packages/rook_sdk_samsung_health/android/src/main/kotlin/io/tryrook/rook_sdk_samsung_health/mapper/PhysicalSummaryMappers.kt
package io.tryrook.rook_sdk_samsung_health.mapper

import io.tryrook.rook_sdk_samsung_health.DEFAULT_DOUBLE
import io.tryrook.rook_sdk_samsung_health.DEFAULT_INT
import io.tryrook.rook_sdk_samsung_health.extension.toProtoString
import io.tryrook.rook_sdk_samsung_health.proto.PhysicalSummaryProto
import io.tryrook.sdk.samsung.domain.model.SHPhysicalSummary

fun SHPhysicalSummary.toProto(): PhysicalSummaryProto {
    return PhysicalSummaryProto.newBuilder()
        .setDateTime(dateTime.toProtoString())
        .setSourceOfData(sourceOfData)
        .setWasTheUserUnderPhysicalActivity(wasTheUserUnderPhysicalActivity)
        .setActiveSeconds(activeSeconds ?: DEFAULT_INT)
        .setRestSeconds(restSeconds ?: DEFAULT_INT)
        .setLowIntensitySeconds(lowIntensitySeconds ?: DEFAULT_INT)
        .setModerateIntensitySeconds(moderateIntensitySeconds ?: DEFAULT_INT)
        .setVigorousIntensitySeconds(vigorousIntensitySeconds ?: DEFAULT_INT)
        .setInactivitySeconds(inactivitySeconds ?: DEFAULT_INT)
        .addAllActivityLevelGranularData(activityLevelGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setContinuousInactivePeriods(continuousInactivePeriods ?: DEFAULT_INT)
        .setCaloriesNetIntakeKcal(caloriesNetIntakeKcal ?: DEFAULT_DOUBLE)
        .setCaloriesExpenditureKcal(caloriesExpenditureKcal ?: DEFAULT_DOUBLE)
        .setCaloriesNetActiveKcal(caloriesNetActiveKcal ?: DEFAULT_DOUBLE)
        .setCaloriesBasalMetabolicRateKcal(caloriesBasalMetabolicRateKcal ?: DEFAULT_DOUBLE)
        .setSteps(steps ?: DEFAULT_INT)
        .addAllStepsGranularData(stepsGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setActiveSteps(activeSteps ?: DEFAULT_INT)
        .addAllActiveStepsGranularData(activeStepsGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setWalkedDistanceMeters(walkedDistanceMeters ?: DEFAULT_DOUBLE)
        .setTraveledDistanceMeters(traveledDistanceMeters ?: DEFAULT_DOUBLE)
        .addAllTraveledDistanceGranularData(
            traveledDistanceGranularData?.map { it.toGranularDataProto() } ?: emptyList(),
        )
        .setFloorsClimbed(floorsClimbed ?: DEFAULT_DOUBLE)
        .addAllFloorsClimbedGranularData(floorsClimbedGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setElevationAvgAltitudeMeters(elevationAvgAltitudeMeters ?: DEFAULT_DOUBLE)
        .setElevationMinimumAltitudeMeters(elevationMinimumAltitudeMeters ?: DEFAULT_DOUBLE)
        .setElevationMaximumAltitudeMeters(elevationMaximumAltitudeMeters ?: DEFAULT_DOUBLE)
        .setElevationLossActualAltitudeMeters(elevationLossActualAltitudeMeters ?: DEFAULT_DOUBLE)
        .setElevationGainActualAltitudeMeters(elevationGainActualAltitudeMeters ?: DEFAULT_DOUBLE)
        .setElevationPlannedGainMeters(elevationPlannedGainMeters ?: DEFAULT_DOUBLE)
        .addAllElevationGranularData(elevationGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setSwimmingNumStrokes(swimmingNumStrokes ?: DEFAULT_INT)
        .setSwimmingNumLaps(swimmingNumLaps ?: DEFAULT_INT)
        .setSwimmingPoolLengthMeters(swimmingPoolLengthMeters ?: DEFAULT_DOUBLE)
        .setSwimmingTotalDistanceMeters(swimmingTotalDistanceMeters ?: DEFAULT_DOUBLE)
        .addAllSwimmingDistanceGranularData(
            swimmingDistanceGranularData?.map { it.toGranularDataProto() } ?: emptyList(),
        )
        .setHrMaximumBpm(hrMaximumBpm ?: DEFAULT_INT)
        .setHrMinimumBpm(hrMinimumBpm ?: DEFAULT_INT)
        .setHrAvgBpm(hrAvgBpm ?: DEFAULT_INT)
        .setHrRestingBpm(hrRestingBpm ?: DEFAULT_INT)
        .addAllHrGranularData(hrGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setHrvAvgRmssd(hrvAvgRmssd ?: DEFAULT_DOUBLE)
        .setHrvAvgSdnn(hrvAvgSdnn ?: DEFAULT_DOUBLE)
        .addAllHrvSdnnGranularData(hrvSdnnGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .addAllHrvRmssdGranularData(hrvRmssdGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setSaturationAvgPercentage(saturationAvgPercentage ?: DEFAULT_DOUBLE)
        .addAllSaturationGranularData(saturationGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setVo2MaxMlPerMinPerKg(vo2MaxMlPerMinPerKg ?: DEFAULT_DOUBLE)
        .addAllVo2GranularData(vo2GranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setStressAtRestDurationSeconds(stressAtRestDurationSeconds ?: DEFAULT_INT)
        .setStressDurationSeconds(stressDurationSeconds ?: DEFAULT_INT)
        .setLowStressDurationSeconds(lowStressDurationSeconds ?: DEFAULT_INT)
        .setMediumStressDurationSeconds(mediumStressDurationSeconds ?: DEFAULT_INT)
        .setHighStressDurationSeconds(highStressDurationSeconds ?: DEFAULT_INT)
        .addAllStressGranularData(stressGranularData?.map { it.toGranularDataProto() } ?: emptyList())
        .setStressAvgLevel(stressAvgLevel ?: DEFAULT_INT)
        .setStressMaximumLevel(stressMaximumLevel ?: DEFAULT_INT)
        .build()
}
